src/interface/swarm_device_context.hpp

```diff
@@ -14,9 +14,28 @@
 #define INTERFACE_SWARM_DEVICE_CONTEXT_HPP_
 
 #include "coordinates/coordinates.hpp"
+#include "utils/utils.hpp"
 
 namespace parthenon {
 
+struct SwarmKey {
+  KOKKOS_INLINE_FUNCTION
+  SwarmKey() {}
+  KOKKOS_INLINE_FUNCTION
+  SwarmKey(const int cell_idx_1d, const int swarm_idx_1d)
+      : cell_idx_1d_(cell_idx_1d), swarm_idx_(swarm_idx_1d) {}
+
+  int cell_idx_1d_;
```

Comment:
Thinking long run here, I suggest to add either a safety count that nx*ny*nz < max int, or increase to int64_t.

Reply:
Yeah this will probably be a problem eventually, I should've been using `int64_t` from the beginning 😬

Reply:
Actually this particular one may always be fine since at most this should be `nx*ny*nz` of a meshblock times the number of meshblocks on a processor. We'd need about 1200^3 cells on a single compute processor to cause an issue here. I'll still add the debug check now just for future-proofing.

I'm actually adding this check where `cell_idx_1d` is evaluated, line 520 of `src/interface/swarm.cpp`, since it will already arrive here as an integer.

Reply:
How about moving the `PARTHENON_DEBUG_REQUIRE` that currently lives inside a parallel region to somewhere outside (e.g., the constructor) and make it a check that is also triggered without debug mode?
My reasoning is that the MeshBlock size won't change for a simulation, so we need the check only once, and that it's rather unlikely to run a problem at scale (where the issue will show first) in debug mode.

PS: I don't think that 1k^3 is that far off. We can already fit multi-stage hydrodynamics at 512^3 on a single A100 with 40G.
So if I take hydro out and just do pure n-body (i.e., reducing the number of cell centered variables needed) and take the next gen GPU with >=128G memory, we may get there sooner than later.

Reply:
Good point about lack of debug mode at scale -- I just moved the `PARTHENON_REQUIRE` up out of the `KOKKOS_LAMBDA` but in the same function so it's clear where the problem is -- I don't think that should hurt performance (and the constructor itself doesn't have a meshblock pointer so the check would have to be in `SetBlockPointer` and then might not be propagated correctly if that part of the code is refactored). 

PS The future is arriving sooner than I expected :)